tests: add test for rebalancer thrashing

cuongdo opened the pull request:
Add a test to ensure that we don't try to rebalance when the cluster is
sufficiently balanced.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9371)

<!-- Reviewable:end -->

Commit:
tests: add test for rebalancer thrashing

Add a test to ensure that we don't try to rebalance when the cluster is
sufficiently balanced.

tamird commented on storage/allocator_test.go:
```diff
@@ -455,6 +457,109 @@ func TestAllocatorRebalance(t *testing.T) {
+	a.options.Deterministic = true
+	defer stopper.Stop()
+
+	stores := []*roachpb.StoreDescriptor{
```
is the fact that there are 4 stores coupled to the fact that there are 4 ranges in the factory functions below? if not, better to make the numbers different to avoid chekhov's gun.

tamird commented on storage/allocator_test.go:
```diff
@@ -455,6 +457,109 @@ func TestAllocatorRebalance(t *testing.T) {
+	// a range count that's above the target range count for the rebalancer, so it
+	// should be rebalanced from.
+	oneReplicaAboveRebalanceTarget := func(mean int32) [4]testReplica {
+		var replicas [4]testReplica
```
perhaps `type testCase [4]testReplica` to avoid repeating yourself.

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/allocator_test.go, line 529 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KRe7d_crucqypd8QPL-:-KRe7d_crucqypd8QPL0:b-rbgf2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d81e9c669aad1743d860d5e8a7962a5d5cda5ef/storage/allocator_test.go#L529)):*
> ```Go
> 		// test the case where 3 nodes are above the mean but below the rebalancer's
> 		// target range count but one node is well below the mean. We still want to
> 		// rebalance to severely underused stores.
> ```

Can you add that scenario here, verify it fails and then comment it out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-:-KRe7pn54vAd0_whmC0i:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/allocator_test.go, line 469 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KRdy9Hq8u-EOy4Ne__c-r1-469:-KReLCasT2LXTGvdXBqk:bbcplbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d81e9c669aad1743d860d5e8a7962a5d5cda5ef/storage/allocator_test.go#L469)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is the fact that there are 4 stores coupled to the fact that there are 4 ranges in the factory functions below? if not, better to make the numbers different to avoid chekhov's gun.
</details>
rewrote substantially to make the intent clearer

---

*[storage/allocator_test.go, line 501 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KRdy9Hq8u-EOy4Ne__c-r1-501:-KReL8cT_JQm_sCF8-KP:b-132xnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d81e9c669aad1743d860d5e8a7962a5d5cda5ef/storage/allocator_test.go#L501)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> perhaps `type testCase [4]testReplica` to avoid repeating yourself.
</details>
no longer relevant

---

*[storage/allocator_test.go, line 529 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KRe7d_crucqypd8QPL-:-KReLA1MT78j667XSF2X:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d81e9c669aad1743d860d5e8a7962a5d5cda5ef/storage/allocator_test.go#L529)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Can you add that scenario here, verify it fails and then comment it out?
</details>
added

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/allocator_test.go, line 475 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLbV6Yu2tZafpkuGB:-KReLbV6Yu2tZafpkuGC:b-q9c620) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L475)):*
> ```Go
> 	// should be rebalanced from.
> 	oneStoreAboveRebalanceTarget := func(mean int32, numStores int) []testStore {
> 		if minStores := 4; numStores < minStores {
> ```

why 4? extract a constant to share with `oneUnderusedStore` and anchor a comment on it.

---

*[storage/allocator_test.go, line 482 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReM6vbqgLkHtGdVIkp:-KReM6vbqgLkHtGdVIkq:bd7lpwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L482)):*
> ```Go
> 			stores[i].rangeCount = mean
> 		}
> 		thresholdDelta := int32(float64(mean)*rebalanceThreshold + 1)
> ```

perhaps `surplus` for consistency with `deficit` below? the math here seems sketchy, too, in that unlike below, the first and last replicas will be exact mirrors. i thought the intention was to have one outlier store.

---

*[storage/allocator_test.go, line 489 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLv4Abhd4xTZB8yXN:-KReLv4Abhd4xTZB8yXO:bbi886q) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L489)):*
> ```Go
> 	}
> 
> 	// Returns a slice of stores with the specified mean (fill in TBD)
> ```

Looks like you forgot to flesh out this comment.

---

*[storage/allocator_test.go, line 494 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLje6bDkdM6jCCc_C:-KReLje6bDkdM6jCCc_D:b-945amf) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L494)):*
> ```Go
> 			t.Fatalf("# of stores %d < min %d", numStores, minStores)
> 		}
> 		stores := make([]testStore, 10)
> ```

s/10/numStores/?

---

*[storage/allocator_test.go, line 498 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLoA5FZ7k_dLoNHuG:-KReLoA5FZ7k_dLoNHuH:b9mr3da) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L498)):*
> ```Go
> 			stores[i].rangeCount = mean
> 		}
> 		// Subtract a large number of ranges from the first store and add that delta
> ```

better if you clarify "large". it seems to be the minimum number that will trigger the rebalancer's heuristic?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/allocator_test.go, line 475 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLbV6Yu2tZafpkuGB:-KReOFxlnfAOpCWQpvBu:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L475)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why 4? extract a constant to share with `oneUnderusedStore` and anchor a comment on it.
</details>
Done

---

*[storage/allocator_test.go, line 482 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReM6vbqgLkHtGdVIkp:-KReP2h-eL92YcbNmTE7:b-szgmkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L482)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> perhaps `surplus` for consistency with `deficit` below? the math here seems sketchy, too, in that unlike below, the first and last replicas will be exact mirrors. i thought the intention was to have one outlier store.
</details>
Renamed. I've made this function more consistent with the other for readability, but functionally, it's a moot point until we implement whatever heuristic we choose to decide if a replica doesn't have enough ranges.

---

*[storage/allocator_test.go, line 489 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLv4Abhd4xTZB8yXN:-KRePW8_hQcY5FObWRjT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L489)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Looks like you forgot to flesh out this comment.
</details>
Done.

---

*[storage/allocator_test.go, line 494 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLje6bDkdM6jCCc_C:-KRePZG4L1q8gFzBTRqX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L494)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/10/numStores/?
</details>
Done.

---

*[storage/allocator_test.go, line 498 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-KReLoA5FZ7k_dLoNHuG:-KReQ4lj4eThr_bSY3J0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08232814434c98258395304d947b389a9980e07b/storage/allocator_test.go#L498)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> better if you clarify "large". it seems to be the minimum number that will trigger the rebalancer's heuristic?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9371#-:-KRepXa0Toqykan5MpCr:bnfp4nl)*
<!-- Sent from Reviewable.io -->
